aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/AWSS3StoragePlugin.java

```diff
@@ -126,20 +127,36 @@ public final class AWSS3StoragePlugin extends StoragePlugin<S3Client> {
 
     private static final int DEFAULT_URL_EXPIRATION_DAYS = 7;
 
-    private final StorageService.Factory storageServiceFactory;
+    private final AWSS3StorageService.Factory storageServiceFactory;
     private final ExecutorService executorService;
-    private final AuthCredentialsProvider authCredentialsProvider;
+    private AuthCredentialsProvider authCredentialsProvider;
     private final AWSS3StoragePluginConfiguration awsS3StoragePluginConfiguration;
     private AWSS3StorageService defaultStorageService;
     @SuppressWarnings("deprecation")
     private StorageAccessLevel defaultAccessLevel;
     private int defaultUrlExpiration;
 
-    private Map<String, AWSS3StorageService> awsS3StorageServicesByBucketName = new HashMap<>();
+    private final Map<String, AWSS3StorageService> awsS3StorageServicesByBucketName = new HashMap<>();
     private Context context;
     @SuppressLint("UnsafeOptInUsageError")
     private List<AmplifyOutputsData.StorageBucket> configuredBuckets;
 
+    @SuppressLint("UnsafeOptInUsageError")
+    private StorageTransferClientProvider clientProvider
+            = new S3StorageTransferClientProvider((region, bucketName) -> {
+                if (region != null && bucketName != null) {
+                    StorageBucket bucket = StorageBucket.fromBucketInfo(new BucketInfo(bucketName, region));
+                    return getAWSS3StorageService((ResolvedStorageBucket) bucket).getClient();
+                }
+
+                if (region != null) {
+                    // unable to create a new S3Client from java code,
+                    // redirecting to AWSS3Service to create S3 Client from kotlin
```

Comment:
i don't quite understand this comment, from the checks it looks like we'll dynamically create the S3 client based on 
1. bucket and region
2. region only- bucket inferred?
3. bucket and region inferred?

what does it mean to "create S3 Client from kotlin"?

Reply:
ideally, we already have the region here in `AWSStoragePlugin.java` and can create an S3Client directly using java syntax i.e. `S3Client.builder().setRegion(region).setCredentialsProvider(authCredentialsProvider).build()`.  But this syntax is not possible from java code since the AWS Kotlin SDK doesn't expose the builder to set the region and credentials.  The only one to create the S3Client is by using kotlin syntax, so the S3Client creation is defined in `AWSS3Service.kt`